### PR TITLE
Update padFootprint function

### PR DESCRIPTION
**Pull request by Nisarg236**
[AB#44902](https://dev.azure.com/rapyuta-robotics/sootballs/_workitems/edit/44902)
Previously this function was scaling the footprint with respect to the coordinate frame 0,0 which could result in distortion. Also when the centroid is having large offset with 0,0 then it will just displace the footprint instead of padding.

Now it scales with respect to the centroid of the footprint.

image
image
another case (10,10), (10,12), (12,12) , (12,10)
image
image

another case (0,0), (0,2), (2,2), (2,0)
image
image

I have also tested the behaviour with negative values of padding (shrinking the footprint)

### CODE REVIEW

**Review by corot — APPROVED** [review]

**Review by corot — CHANGES_REQUESTED** [review]
3 things

The 1st example looks suspicious:  doesn't look like padded 10 cm but 2 or 3  :thinking: 

This looks similar to [this commit](https://github.com/corot/navigation/commit/7c4fbb9338178a240425f9b93efa994b0bb0f00d)
Do both provide the same result?  If so, this is simpler

Keep the formatting of the files you modify. i.e. never mix stiles

**Comment by Nisarg236** [comment]
> 3 things
> 
> The 1st example looks suspicious: doesn't look like padded 10 cm but 2 or 3 🤔
> 
> This looks similar to [this commit](https://github.com/corot/navigation/commit/7c4fbb9338178a240425f9b93efa994b0bb0f00d) Do both provide the same result? If so, this is simpler
> 
> Keep the formatting of the files you modify. i.e. never mix stiles

Hi you are right it wasnt 0.1 meters because it was scaling by percentage, I changed it to scale by meters and updated the image.
and now the functionality is the same as the one you shared.

**Comment by corot** [comment]
> this commit

**Comment by corot** [comment]
> Hi you are right it wasnt 0.1 meters because it was scaling by percentage, I changed it to scale by meters and updated the image. and now the functionality is the same as the one you shared.

So the outcome is now the same as with [this commit](https://github.com/corot/navigation/commit/7c4fbb9338178a240425f9b93efa994b0bb0f00d)?  
Then better to use the commit,,, as the logic is much simpler  :thinking:

**Comment by renan028** [comment]
> > Hi you are right it wasnt 0.1 meters because it was scaling by percentage, I changed it to scale by meters and updated the image. and now the functionality is the same as the one you shared.
> 
> So the outcome is now the same as with [this commit](https://github.com/corot/navigation/commit/7c4fbb9338178a240425f9b93efa994b0bb0f00d)? Then better to use the commit,,, as the logic is much simpler 🤔

@corot I think you are missing the part he calculates the centroid of the polygon.

File: costmap_2d/src/footprint.cpp [hunk 1]
```diff
@@ -137,12 +137,29 @@ void transformFootprint(double x, double y, double theta, const std::vector<geom
...
+  geometry_msgs::Point footprint_center;
+  double sum_x = 0.0, sum_y = 0.0;
+
+  for (const auto& point : footprint)
+  {
+    sum_x += point.x;
+    sum_y += point.y;
+  }
+  footprint_center.x = sum_x / footprint.size();
+  footprint_center.y = sum_y / footprint.size();
```
Comment (renan028) [inline costmap_2d/src/footprint.cpp]: S: I think it makes sense to create a method `centroid` that returns `geometry_msgs::Point`

File: costmap_2d/src/footprint.cpp [hunk 2]
```diff
@@ -137,12 +137,29 @@ void transformFootprint(double x, double y, double theta, const std::vector<geom
...
+
+  for (auto& point : footprint)
   {
-    geometry_msgs::Point& pt = footprint[ i ];
-    pt.x += sign0(pt.x) * padding;
-    pt.y += sign0(pt.y) * padding;
+    double dx = point.x - footprint_center.x;
+    double dy = point.y - footprint_center.y;
+
+    double distance = std::sqrt(dx * dx + dy * dy);
```
Comment (renan028) [inline costmap_2d/src/footprint.cpp]: S: `std::hypot`

**Review by renan028 — APPROVED** [review]

**Comment by renan028** [comment]
I think it makes sense to make a PR upstream:
https://github.com/ros-navigation/navigation2/blob/4e5d2dfc66cb75eb390d614d13e5b64efbf30284/nav2_costmap_2d/src/footprint.cpp#L145

File: costmap_2d/test/footprint_tests.cpp [hunk 3]
```diff
@@ -70,17 +70,17 @@ TEST( Costmap2DROS, padded_footprint_from_string_param )
   std::vector<geometry_msgs::Point> footprint = cm.getRobotFootprint();
   EXPECT_EQ( 3, footprint.size() );
 
-  EXPECT_EQ( 1.5f, footprint[ 0 ].x );
-  EXPECT_EQ( 1.5f, footprint[ 0 ].y );
-  EXPECT_EQ( 0.0f, footprint[ 0 ].z );
+  EXPECT_EQ( 1.66667f, footprint[0].x );
```
Comment (corot) [inline costmap_2d/test/footprint_tests.cpp]: Q: do we really need to change round numbers?

**Review by corot — APPROVED** [review]
ok, as @renan028 explained, this implementation is superior to the one I mention

@Nisarg236 , you should consider open a PR to nav2, as this is a nice contribution

File: costmap_2d/test/footprint_tests.cpp (hunk 3 above)
Comment (Nisarg236) [reply costmap_2d/test/footprint_tests.cpp]: yes because otherwise it fails
